Earn: Move donations block to Jetpack Beta Blocks

gwwar opened the pull request:
As discussed in p1HpG7-9Kg-p2#comment-40839, we're moving development of the Donation block to Jetpack from the FSE plugin. This block will allow customers on Jetpack and WordPress.com to accept one-time or recurring monthly/annual donations.

| Site | No Plan | No Stripe Connection | Edit View |
| --- | --- | --- | --- |
| Jetpack | N/A. Block cannot be inserted in JP sites under a free plan. This is by design because we're using the `plan_check` property during block registration which makes the block non-insertable unless the `jetpack_block_editor_enable_upgrade_nudge` filter is enabled (it's disabled by default on JP sites). | <img width="400" alt="Screen Shot 2020-07-23 at 14 03 25" src="https://user-images.githubusercontent.com/1233880/88284218-5397e380-cced-11ea-8669-0b4dba67e382.png"> | <img width="400" alt="Screen Shot 2020-07-23 at 14 04 35" src="https://user-images.githubusercontent.com/1233880/88284294-775b2980-cced-11ea-8c6d-cd638ae57ad3.png"> |
| Simple | <img width="400" alt="Screen Shot 2020-07-23 at 14 18 04" src="https://user-images.githubusercontent.com/1233880/88285438-5bf11e00-ccef-11ea-9293-b606e5ee8675.png"> | <img width="400" alt="Screen Shot 2020-07-23 at 14 19 17" src="https://user-images.githubusercontent.com/1233880/88285521-84791800-ccef-11ea-8847-3e69f5e8fef5.png"> | <img width="400" alt="Screen Shot 2020-07-23 at 14 20 14" src="https://user-images.githubusercontent.com/1233880/88285580-a7a3c780-ccef-11ea-9213-971d0743ec6c.png"> |
| Atomic | N/A. All Atomic sites are under a paid plan. (except for downgraded sites) | <img width="1166" alt="Screen Shot 2020-07-23 at 11 17 20 AM" src="https://user-images.githubusercontent.com/1270189/88324940-47a22680-ccd9-11ea-9ecf-7967c3ed2f53.png"> | <img width="1099" alt="Screen Shot 2020-07-23 at 11 29 13 AM" src="https://user-images.githubusercontent.com/1270189/88324893-3527ed00-ccd9-11ea-9f30-784abb2bf66b.png"> |

 This PR builds on top of work in:
* https://github.com/Automattic/wp-calypso/pull/43798
* https://github.com/Automattic/wp-calypso/pull/43834
* https://github.com/Automattic/wp-calypso/pull/43166
* https://github.com/Automattic/wp-calypso/pull/43665
* https://github.com/Automattic/wp-calypso/pull/42911

### What @gwwar and @mmtr changed from copied FSE files
- Block now has Jetpack namespace, eg `jetpack/donations`
- Text domain updated from `full-site-editing` to `jetpack`
- editor styles namespaced with generated block name `wp-block-jetpack-donations`
- do not fetch default payment products when user does not have a stripe connection or correct paid plan. (this would previously cause the API error to override the stripe/plan nudges).
- do not remove loader until default products are available, since block options were non-interactive. (Block attributes were unresponsive without this, but this is kind of slow so I might back this out.)
- This block has been added to the Jetpack Beta list, meaning that this won't be visible to folks unless `define( 'JETPACK_BETA_BLOCKS', true )` is set on your WordPress instance
- Avoid new Gutenberg features eg `import { __experimentalBlock as Block } from '@wordpress/block-editor';`

### What we're not including in this PR
We still need to pick final **serialization options for the published block**, but this can be done in a follow up PR. With Donations now in Jetpack it's not unreasonable to nest a child payments button, or instead opt to go with something a bit more custom but duplicate some of the iframe handling. Other pieces like the Edit view and backend glue pieces needed for the actual transactions are mostly done, but may need some follow up

## Testing Instructions

### Jetpack sites
- With a WordPress instance of your choice, install Jetpack Beta and install this branch. Or visit jurassic.ninja/create/?jetpack-beta&branch=add/donation-block-scaffolding
- Connect Jetpack to WordPress.com with a free plan
- Make sure Jetpack Blocks are visible by clicking on /wp-admin/options-general.php?page=companion_settings and selecting 'JETPACK_BETA_BLOCKS'
<img width="1063" alt="Screen Shot 2020-07-21 at 5 21 00 PM" src="https://user-images.githubusercontent.com/1270189/88119896-91bec700-cb76-11ea-8f20-373fb51118db.png">

- Try inserting the donations block. The block shouldn't appear in the block inserter.
- Purchase a personal plan or higher. Note that if folks are using the sandbox store method (PCYsg-lW4-p2) to also set the JETPACK__SANDBOX_DOMAIN value.
- Refresh the editor and insert a Donations block. We should now see a nudge to connect stripe. The block should be fully editable, although behind the scenes we're using fake plans since there is no connection to Stripe (you can check this by switching to the code editor and noting plan IDs are set with a -1 value).
- Create a real stripe account, or use a dev account noted in "Sandbox method" PCYsg-lW4-p2
- Refresh the editor again. Make sure the interactive edit view for donation blocks allows us to modify the types of donations (one-time/monthly/yearly), and lets us edit the text fields. Switch to the code editor and confirm the plan IDs attributes have been updated to the real plans.

### Simple sites
- Apply D46803-code and D46705-code.
- Try inserting the donations block. The block should appear in the block inserter with a paid flag. Once inserted, we should see a nudge to upgrade. The block should be fully editable, although behind the scenes we're using fake plans since there is no connection to Stripe (you can check this by switching to the code editor and noting plan IDs are set with a -1 value).
- Purchase a personal plan or higher.
- Refresh the editor. We should now see a nudge to connect stripe. The block should continue to be fully editable (and keep using the fake plans).
- Create a real stripe account, or use a dev account noted in "Sandbox method" PCYsg-lW4-p2
- Refresh the editor again. Make sure the interactive edit view for donation blocks allows us to modify the types of donations (one-time/monthly/yearly), and lets us edit the text fields. Switch to the code editor and confirm the plan IDs attributes have been updated to the real plans.

### Atomic Sites
- Apply D46803-code and sandbox
- Setup a real atomic site by purchasing a Business plan on a WordPress.com account and transferring by installing a custom plugin or theme
- Install the Jetpack Beta plugin https://jetpack.com/download-jetpack-beta/
- Activate this branch `add/donation-block-scaffolding`
- Set `add_filter( 'jetpack_blocks_variation', function() { return 'beta'; } );` to activate beta blocks. A quick way of doing this is modifying a simple plugin with the plugin editor, like hello dolly.
- In the block-editor, Try adding a donations block, we should see a stripe connection prompt
- Click on the connect stripe nudge, and create a real stripe account
- In the block-editor try adding a donations block, we should see an editable donations view, and real planIds set in the block attributes of code view.

gwwar commented:
@Automattic/serenity feel free to hop onto this branch, just let folks know if y'all are planning to rebase.

gwwar commented on modules/module-extras.php:
```diff
@@ -44,6 +44,7 @@
 // Some features are only available when connected to WordPress.com.
 $connected_tools = array(
 	'calypsoify/class.jetpack-calypsoify.php',
+	'memberships/class-jetpack-memberships.php',
```
Good catch @mmtr

kwight commented:
Fired this up quickly to check regressions with the Payments block – just noting a few things for later sake, understanding it's plumbing that's the focus of this PR.

Applying the branch through Jetpack Beta on an existing JN site caused fatals, but doing a clean JN install with the link in the instructions worked fine:

<img width="925" alt="Screen Shot 2020-07-22 at 10 50 37 AM" src="https://user-images.githubusercontent.com/349751/88211979-5b7f5700-cc0b-11ea-90a3-64d077513f44.png">

The Donations block displays its "read more" link in line with the Upgrade button, rather than below it like the Payments button.

<img width="699" alt="Screen Shot 2020-07-22 at 10 58 27 AM" src="https://user-images.githubusercontent.com/349751/88212304-e5c7bb00-cc0b-11ea-8e92-556a81bdb6ec.png">
<img width="698" alt="Screen Shot 2020-07-22 at 11 09 40 AM" src="https://user-images.githubusercontent.com/349751/88212306-e8c2ab80-cc0b-11ea-9e3a-7270b6a53fd7.png">

Trying to use the Payments block on the front end, I had the same experience as @gwwar : the button would resolve to the post URL with a hash (`#`) on the end, with no obvious errors at all.

gwwar commented:
@mmtr I ended up needing to back out the membership registration refactors. Tracing the code we were getting to the correct register_jetpack_blocks calls, but the render_callback for the payment buttons did not trigger. Let's try this improvement out in smaller a follow up PR

gwwar commented:
Alright @Automattic/serenity I think this one is good to test, since this is already rather large. I'll add a few follow up issues.

mmtr commented:
> @mmtr I ended up needing to back out the membership registration refactors. Tracing the code we were getting to the correct register_jetpack_blocks calls, but the render_callback for the payment buttons did not trigger. Let's try this improvement out in smaller a follow up PR

Thanks for reverting my changes! Main reason why I did refactor the registration of the Payments block is because I wasn't able to see the Donations block, but I think I was missing to activate the beta blocks (I wrongly assumed they are enabled by default on a dev environment), because I can see it know.

mmtr commented on extensions/blocks/donations/donations.php:
```diff
@@ -0,0 +1,49 @@
+		array( 'render_callback' => __NAMESPACE__ . '\load_assets' )
+	);
+}
+add_action( 'init', __NAMESPACE__ . '\register_block' );
```
I think we usually hook into `jetpack_register_gutenberg_extensions` for registering the Jetpack blocks. Will look into it.

jeherve replied:
`init` works here 👍

gwwar replied:
I used the `wp jetpack scaffold` command here to generate the init block boilerplate, but let us know if there's a preferred style.

jeherve commented on extensions/blocks/donations/attributes.js:
```diff
@@ -0,0 +1,64 @@
+/**
+ * WordPress dependencies
+ */
+// eslint-disable-next-line wpcalypso/import-docblock
```
I think this (and the ones in other files) can be removed now.

mmtr replied:
Unfortunately we still need it, since JP eslint extends `wpcaypso/react`:

https://github.com/Automattic/jetpack/blob/96d6172619dc7caca1c95d617f4fe5ac766c0158/.eslintrc.js#L4

<img width="733" alt="Screen Shot 2020-07-23 at 11 03 36" src="https://user-images.githubusercontent.com/1233880/88269376-34d92300-ccd4-11ea-84ea-f25a7fd06243.png">

jeherve replied:
Ah, right, I hadn't noticed what that rule was about, my bad!

That's a bit of an odd rule, I wonder if it could be improved so it doesn't trigger when there are no external or internal dependencies to import. I opened https://github.com/Automattic/wp-calypso/issues/44387 and https://github.com/Automattic/wp-calypso/pull/44388 to discuss this.

jeherve replied:
We can now remove the rule exceptions: #16566.

jeherve commented on extensions/blocks/donations/donations.php:
```diff
@@ -0,0 +1,49 @@
+ * registration if we need to.
+ */
+function register_block() {
+	jetpack_register_block(
+		BLOCK_NAME,
+		array( 'render_callback' => __NAMESPACE__ . '\load_assets' )
+	);
```
Is the block going to be available to all users, or are you planning on restricting it to some specific plans? If the latter, you'd want to add a plan check here:
https://github.com/Automattic/jetpack/blob/5a4ec0da4f32874708b4f7014690aa458ceca883/extensions/README.md#L157

mmtr replied:
We want the block to be available in all plans, but usable only on paid plans. Users on a free plan can insert the block but they will see an upgrade nudge:

https://github.com/Automattic/jetpack/blob/1a8ca1cd66aaee9b711e1bb15a65ee7b113a5a84/extensions/blocks/donations/edit.js#L96-L98

Due to that, we want the block to be always registered. AFAIK we cannot use the `plan_check => true` arg because that will hide the block from the inserter. The `jetpack_block_editor_enable_upgrade_nudge` filter could help displaying the block in the inserter, but seems it doesn't currently offer enough granularity so we can display only this block without affecting the rest of paid blocks that are not currently shown in the inserter.

jeherve commented on extensions/blocks/donations/editor.js:
```diff
@@ -0,0 +1,47 @@
+import registerJetpackBlock from '../../shared/register-jetpack-block';
+import { name, settings } from '.';
+
+/**
+ * Appends a "paid" tag to the Donations block title if site requires an upgrade.
+ */
+const addPaidBlockFlags = async () => {
```
I would recommend relying on the existing Jetpack tools to handle this. You can look at the Calendly check for an example, or the Pay With PayPal block.

mmtr replied:
Do we have any existing issue with the paid labels? I don't see the flag on the Calendly block after setting a JN sites on a free plan:

<img width="423" alt="Screen Shot 2020-07-23 at 10 48 36" src="https://user-images.githubusercontent.com/1233880/88268119-3ace0480-ccd2-11ea-9e71-1d79d367f05b.png">

jeherve replied:
The Calendly block is free for self-hosted users; we only apply the plan restriction on WordPress.com Simple sites.

jeherve commented on extensions/blocks/donations/editor.scss:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Editor styles for Donations
+ */
+.wp-block-jetpack-donations .donations__container {
```
I'd recommend switching from spaces to tabs, to ensure that the next person editing this file does not fix all that at once.

jeherve commented on extensions/blocks/donations/editor.scss:
```diff
@@ -0,0 +1,72 @@
+      padding: 16px;
+      height: auto;
+      border-radius: 0;
+      border-left: 1px solid #a7aaad;
```
I would recommend relying on core colors for things like this:
https://github.com/Automattic/jetpack/blob/5a4ec0da4f32874708b4f7014690aa458ceca883/extensions/README.md#L279-L285

jeherve commented on extensions/blocks/donations/stripe-nudge.js:
```diff
@@ -0,0 +1,56 @@
+import { __ } from '@wordpress/i18n';
+import { withDispatch } from '@wordpress/data';
+
+const StripeNudge = ( { autosaveAndRedirect, stripeConnectUrl } ) => (
```
Could we rely on the existing component for this?
https://github.com/Automattic/jetpack/blob/master/extensions/shared/components/stripe-nudge/index.jsx

jeherve commented on extensions/blocks/donations/tab.js:
```diff
@@ -0,0 +1,122 @@
+ */
+import Context from './context';
+import formatCurrency, { CURRENCIES } from '@automattic/format-currency';
+import { minimumTransactionAmountForCurrency } from '../recurring-payments';
```
If we're re-using this across more than one block, I think it would make sense to move it to `extensions/shared`.

jeherve commented on extensions/blocks/donations/upgrade-plan.js:
```diff
@@ -0,0 +1,38 @@
+/* eslint-disable wpcalypso/import-docblock */
```
We may be able to move away from this if we switch to relying on existing tools to mark the block as paid. When we do so, we'll benefit from `wrapPaidBlock` out of the box.

mmtr commented:
Thanks for the feedback @jeherve. I updated the PR with your suggestions and we now register the block using the `plan_check` property so we can reuse the available infrastructure in Jetpack for registering paid blocks.

Is there any way we can enable the beta blocks in Atomic sites? Seems they are disabled on AT sites according to PCYsg-iiu-p2.

gwwar commented:
> Is there any way we can enable the beta blocks in Atomic sites?

@mmtr I think, an ephemeral testing site is probably the closest env if there's an explicit check for `IS_ATOMIC`. Or we can temporarily add the block to the production list locally for testing.

jeherve commented:
> Is there any way we can enable the beta blocks in Atomic sites?

You get Experimental blocks there by default (they're loaded very early in an mu-plugin), but you can overwrite that setting like so:

```php
add_filter( 'jetpack_blocks_variation', function() { return 'beta'; } );
```

gwwar commented:
Thanks @jeherve !

jeherve commented on extensions/blocks/donations/editor.scss:
```diff
@@ -0,0 +1,92 @@
+			margin-bottom: 30px;
+		}
+	}
+
+	.jetpack-block-nudge {
+		max-width: none;
+	}
```
I should we should be able to remove this now?

```suggestion

```

gwwar commented:
Reserving this comment for testing:

Jetpack Free Plan:
- We may want to check in with Ajax and @lessbloat on paid nudge work. I think behavior here is fine for free (not available in inserter for now, but we should align Payments/Donations together with whatever we pick).
<img width="774" alt="Screen Shot 2020-07-23 at 10 44 06 AM" src="https://user-images.githubusercontent.com/1270189/88319795-83d18900-ccd1-11ea-8c12-b9d3081960d3.png">

jeherve commented:
>  we should align Payments/Donations together with whatever we pick

Yes, the Payments block was developed prior our work on upgrade nudges, so it's a bit different. This is tracked in #14893.

gwwar commented:
> We've moved to a Button inner block for most of our blocks including buttons. Here is the Payments block, for instance: #15915. Should we do the same for this block?

That sounds reasonable to me @jeherve. Do you prefer if we added that on top of this PR or in a follow up? We haven't picked our final block serialization yet, so there's a lot of flexibility here to play with the edit/save views.

jeherve commented:
> Do you prefer if we added that on top of this PR or in a follow up?

I think it would be fine here, but if y'all would rather ship that first Beta version of the block on WordPress.com quickly, we can do it in a follow-up PR. Up to you, I'm good either way!

gwwar commented:
If you don't mind @jeherve, we can follow up immediately with the Button update in the next PR. This one is a good size, and we should more easily be able to keep track of reasoning for future searchers.

kwight commented:
I ran through this with both the Donations and Payments blocks, on both Simple and JP sites, didn't see anything unexpected 🎉 

(I didn't expect to see the Donations block in the inserter at all without a plan, and didn't expect any front-end output for Donations yet either.)

gwwar commented:
Thanks for the extensive review @jeherve !

I can't accept my own PR, but I did verify that improvements added by @mmtr work well and don't regress the payments button. 💖 Was there any other feedback we should address in this PR?

The team is planning to immediately follow up on the Button improvement and serialization options in the next PR.

mmtr commented:
Thanks for the review @jeherve! Do you why merging is still blocked by the "Required review" check even after your approval?
